Simplify compiler flags checking in Source/ThirdParty/skia/CMakeLists.txt

aperezdc opened the pull request:
#### e739d6576a590c7211f9438c5fa9d54fd4cb1ee6
<pre>
Simplify compiler flags checking in Source/ThirdParty/skia/CMakeLists.txt
<a href="https://bugs.webkit.org/show_bug.cgi?id=270162">https://bugs.webkit.org/show_bug.cgi?id=270162</a>

Reviewed by Michael Catanzaro.

Add new WEBKIT_CHECK_COMPILER_FLAGS and WEBKIT_ADD_COMPILER_FLAGS
functions, which split the existing functionality of the existing
WEBKIT_ADD_TARGET_{C,CXX}_FLAGS macros, making them more generic.
The old macros are kept for now, but implemented in terms of the
new functions.

While at it, the WEBKIT_ADD_COMPILER_FLAGS function was augmented to
support applying flags to SOURCE files and DIRECTORY paths, and a
tweak added to ensure results of checks for flags containing equals
characters (=) are properly cached by CMake to save some time in CMake
re-runs.

* Source/ThirdParty/skia/CMakeLists.txt: Make use of the new
  WEBKIT_CHECK_COMPILER_FLAGS and WEBKIT_ADD_COMPILER_FLAGS functions.
* Source/cmake/WebKitCompilerFlags.cmake: Split and improve
  functionality from the WEBKIT_ADD_TARGET_{C,CXX}_FLAGS macros into
  a new function to check flags and another generic function to apply
  flags to targets, sources, or directories.

Canonical link: <a href="https://commits.webkit.org/275430@main">https://commits.webkit.org/275430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a6c84d7dcdbf3aca511a56682de43b0b3ae0167

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41716 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20730 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44097 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44290 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37809 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18061 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34473 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35923 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15145 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15347 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36945 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45685 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/35169 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37899 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37266 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41024 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/41342 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16531 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13569 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39456 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18150 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/48353 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18207 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9850 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5602 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17794 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (approved):
Interesting.

I like how you used function() instead of macro() to ensure _cachevar doesn't need to be manually unset.

aperezdc committed:
Simplify compiler flags checking in Source/ThirdParty/skia/CMakeLists.txt
https://bugs.webkit.org/show_bug.cgi?id=270162

Reviewed by Michael Catanzaro.

Add new WEBKIT_CHECK_COMPILER_FLAGS and WEBKIT_ADD_COMPILER_FLAGS
functions, which split the existing functionality of the existing
WEBKIT_ADD_TARGET_{C,CXX}_FLAGS macros, making them more generic.
The old macros are kept for now, but implemented in terms of the
new functions.

While at it, the WEBKIT_ADD_COMPILER_FLAGS function was augmented to
support applying flags to SOURCE files and DIRECTORY paths, and a
tweak added to ensure results of checks for flags containing equals
characters (=) are properly cached by CMake to save some time in CMake
re-runs.

* Source/ThirdParty/skia/CMakeLists.txt: Make use of the new
  WEBKIT_CHECK_COMPILER_FLAGS and WEBKIT_ADD_COMPILER_FLAGS functions.
* Source/cmake/WebKitCompilerFlags.cmake: Split and improve
  functionality from the WEBKIT_ADD_TARGET_{C,CXX}_FLAGS macros into
  a new function to check flags and another generic function to apply
  flags to targets, sources, or directories.

Canonical link: https://commits.webkit.org/275430@main
